### PR TITLE
fix: all fields required and mandatory fields when complete

### PR DESCRIPTION
**Pull request by mmmateos**
BREAKING CHANGES: Migrate DataValuePresenter to Kotlin
Check all fields required and mandatory fields when completing dataSet
Jira Issues:
https://jira.dhis2.org/browse/ANDROAPP-2470
https://jira.dhis2.org/browse/ANDROAPP-2468

### CODE REVIEW

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+                            table.indexOf(row),
+                            true
+                        )
+                    }
+                }
+            }
+        }
+        return mandatoryOk
+    }
+
```
Comment (mmmateos) [inline app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: These are the modified methods

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+                            List<DataSetTableModel>,
+                            List<DataElementOperand>,
+                            List<DataElementOperand>,
+                            Sextet<CategoryCombo,
+                                List<DataElement>,
+                                Map<String, List<List<Pair<CategoryOption, Category>>>>,
+                                List<DataSetTableModel>,
+                                List<DataElementOperand>,
+                                List<DataElementOperand>>>
+                        { val0, val1, val2, val3, val4, val5 ->
```
Comment (Balcan) [inline app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: val0, val1...val5 I don't know what are these variables.
Not for now but we will have to check this together :)

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+                    )
+
+                    setTableData(dataTableModel!!)
+                }
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .subscribe(
+                    { quartet ->
+                        view.setTableData(
+                            quartet.val0(),
```
Comment (Balcan) [inline app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: in kotlin you can add a variable name so you could do something like:
variableName1 = quartet.val0(),
variableName2 = quartet.val1(),
...

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+        val isEditable = accessDataWrite &&
+            !isExpired(dataSet) &&
+            dataInputPeriodModel.isEmpty() || (
+            checkHasInputPeriod() != null && DateUtils.getInstance().isInsideInputPeriod(
+                checkHasInputPeriod()
+            )
+            ) &&
+            !isApproval
+
+        return Quartet.create(dataTableModel, listFields, cells, isEditable)
```
Comment (Balcan) [inline app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: Not for now. But this method is way too long (155 lines).
Would be better to split it into several smaller methods.

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt (hunk 1 above)
Comment (Balcan) [reply app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: Just so the others reviewers know:

- checkAllFieldRequired
- checkMandatoryField

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt (hunk 3 above)
Comment (ferdyrod) [reply app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: Also, in kotlin we can deconstruct an object
    ` val (varName1, varName2, varNmae3, varName4) = quartet`

and varName1 will have quartet.val0() value and so forth.

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt (hunk 1 above)
Comment (JaimeToca) [reply app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: any unit test?

File: app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt (hunk 4 above)
Comment (JaimeToca) [reply app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt]: or classes I see too many responsabilities in same method and things that perhaps the presenter should not do. Let's take into account next time when performing unit tests after the release.

**Pull request merged**